server/src/main/kotlin/delta/codecharacter/server/user/public_user/PublicUserService.kt

```diff
@@ -71,11 +69,50 @@ class PublicUserService(@Autowired private val publicUserRepository: PublicUserR
                     wins = it.wins,
                     losses = it.losses,
                     ties = it.ties
-                )
+                ),
             )
         }
     }
 
+    fun updateLeaderboard(leaderboardList: List<LeaderboardEntryDto>) {
+        val leaderboardSize = leaderboardList.size
+        for (i in 0 until leaderboardSize) {
+            val user = publicUserRepository.findByUsername(leaderboardList[i].user.username).get()
+            if (i < 0.2 * leaderboardSize) {
+                publicUserRepository.save(user.copy(tier = TierTypeDto.TIER1))
+            } else {
+                publicUserRepository.save(user.copy(tier = TierTypeDto.TIER2))
+            }
+        }
+    }
+
+    fun updateTempLeaderboardByTier() {
+        updateLeaderboard(getLeaderboardList())
+    }
+
+    fun updateLeaderboard() {
+        val leaderboardList = getLeaderboardList().sortedBy { it.stats.rating }.reversed()
+        updateLeaderboard(leaderboardList)
+    }
+
+    fun getLeaderboardByTier(tier: TierTypeDto?): List<LeaderboardEntryDto> {
+        val leaderboardList = getLeaderboardList().sortedBy { it.stats.rating }.reversed()
+        val leaderboardSize = leaderboardList.size
+        return if (tier == TierTypeDto.TIER1) {
+            leaderboardList.subList(0, (0.2 * leaderboardSize).toInt())
+        } else {
+            leaderboardList.subList((0.2 * leaderboardSize).toInt(), leaderboardList.size)
+        }
+    }
+
+    fun getLeaderboard(page: Int?, size: Int?, tier: TierTypeDto?): List<LeaderboardEntryDto> {
+        val leaderboardEntry = getLeaderboardByTier(tier)
+        if (size!! > leaderboardEntry.size) {
+            return leaderboardEntry
+        }
+        return leaderboardEntry.subList(page!! * size, (page * size) + size)
```

Comment:
- What exactly is happening here?. For pagination we have already used PageRequest in it.
-  Make use of some query functions like `findAllByTier` for querying by `tier`.
- Avoid using `!!`.